### ISSUE TITLE
[ologram] Cython multiprocessing error

### ISSUE THREAD

**Issue by qferre** [dputhier/pygtftk#67]
Weird multiprocessing error when working on certain BED files. Investigating.

```
DeprecationWarning: 'source deactivate' is deprecated. Use 'conda deactivate'.
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 16
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	peak_anno
	1

[Wed Mar 13 15:55:33 2019]
rule peak_anno:
    input: /gpfs/tagc/home/sadouni/tf/split_tf/Ahr::Arnt.bed, input/silencer/pPGK_silencer_R1.bed, soft/pygtftk/done
    output: output/pPGK_silencer_R1/Ahr::Arnt/00_peak_anno_stats.tsv
    jobid: 0
    wildcards: sil=pPGK_silencer_R1, tf=Ahr::Arnt
    threads: 14

 |-- 15:58-INFO-peak_anno : Checking chromosome info file.
 |-- 15:58-INFO-peak_anno : Processing user-defined regions (bed format).
 |-- 15:58-INFO-peak_anno : Processing Ahr__Arnt
 |-- 16:01-INFO-peak_anno : Beginning shuffling for a given set of features...
/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/site-packages/pybedtools/bedtool.py:3439: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\t'.
  return pandas.read_table(self.fn, *args, **kwargs)
 |-- 16:08-INFO-peak_anno : Warning - You have a bed file with features after the end of chromosome "chr10" !
 |-- 16:09-INFO-peak_anno : Warning - You have a bed file with features after the end of chromosome "chr11" !
 |-- 16:11-INFO-peak_anno : Warning - You have a bed file with features after the end of chromosome "chr13" !
 |-- 16:13-INFO-peak_anno : Warning - You have a bed file with features after the end of chromosome "chr15" !
 |-- 16:19-INFO-peak_anno : Warning - You have a bed file with features after the end of chromosome "chr2" !
 |-- 16:20-INFO-peak_anno : Warning - You have a bed file with features after the end of chromosome "chr3" !
 |-- 16:21-INFO-peak_anno : Warning - You have a bed file with features after the end of chromosome "chr4" !
 |-- 16:24-INFO-peak_anno : Warning - You have a bed file with features after the end of chromosome "chr6" !
 |-- 16:28-INFO-peak_anno : Warning - You have a bed file with features after the end of chromosome "chr9" !
 |-- 16:30-INFO-peak_anno : Warning - You have a bed file with features after the end of chromosome "chrX" !
 |-- 16:30-INFO-peak_anno : Warning - You have a bed file with features after the end of chromosome "chrY" !
Traceback (most recent call last):
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/bin/gtftk", line 4, in <module>
    __import__('pkg_resources').run_script('pygtftk==0.9.10.dev0+1a74', 'gtftk')
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/site-packages/pkg_resources/__init__.py", line 666, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1446, in run_script
    exec(code, namespace, namespace)
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/site-packages/pygtftk-0.9.10.dev0+1a74-py3.6-linux-x86_64.egg/EGG-INFO/scripts/gtftk", line 104, in <module>
    args = main()
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/site-packages/pygtftk-0.9.10.dev0+1a74-py3.6-linux-x86_64.egg/EGG-INFO/scripts/gtftk", line 89, in main
    CmdManager.run(args)
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/site-packages/pygtftk-0.9.10.dev0+1a74-py3.6-linux-x86_64.egg/pygtftk/cmd_manager.py", line 947, in run
    fun(**args)
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/site-packages/pygtftk-0.9.10.dev0+1a74-py3.6-linux-x86_64.egg/pygtftk/plugins/peak_anno.py", line 613, in peak_anno
    bedB=BedTool(bed_anno.name))
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/site-packages/pygtftk-0.9.10.dev0+1a74-py3.6-linux-x86_64.egg/pygtftk/stats/intersect/overlap_stats_shuffling.py", line 189, in compute_overlap_stats
    use_markov_shuffling, nb_threads)
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/site-packages/pygtftk-0.9.10.dev0+1a74-py3.6-linux-x86_64.egg/pygtftk/stats/intersect/overlap_stats_shuffling.py", line 81, in compute_all_intersections_minibatch
    shuffled_Li2_batches)
  File "pygtftk/stats/intersect/create_shuffles.pyx", line 276, in pygtftk.stats.intersect.create_shuffles.batch_to_bedlist
  File "pygtftk/stats/intersect/create_shuffles.pyx", line 277, in pygtftk.stats.intersect.create_shuffles.batch_to_bedlist
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/multiprocessing/pool.py", line 288, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/multiprocessing/pool.py", line 670, in get
    raise self._value
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/multiprocessing/pool.py", line 450, in _handle_tasks
    put(task)
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/pygtftk/lib/python3.6/multiprocessing/connection.py", line 393, in _send_bytes
    header = struct.pack("!i", n)
struct.error: 'i' format requires -2147483648 <= number <= 2147483647
[Wed Mar 13 16:33:23 2019]
Error in rule peak_anno:
    jobid: 0
    output: output/pPGK_silencer_R1/Ahr::Arnt/00_peak_anno_stats.tsv

RuleException:
CalledProcessError in line 60 of /gpfs/tagc/home/sadouni/peak_anno_test/snake_file.py:
Command 'set -euo pipefail;  set +u; source /gpfs/tagc/home/sadouni/miniconda3/bin/activate pygtftk; set -u
    gtftk peak_anno -i toto.gtf --no-basic-feature -V 1 -k 14 -p input/silencer/pPGK_silencer_R1.bed -c mm9_chr_size.txt --more-bed /gpfs/tagc/home/sadouni/tf/split_tf/Ahr::Arnt.bed --no-date  -o output/pPGK_silencer_R1/Ahr::Arnt --more-bed-labels Ahr__Arnt' returned non-zero exit status 1.
  File "/gpfs/tagc/home/sadouni/peak_anno_test/snake_file.py", line 60, in __rule_peak_anno
  File "/gpfs/tagc/home/sadouni/miniconda3/envs/snakemake/lib/python3.6/concurrent/futures/thread.py", line 56, in run
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message

```

**Comment by qferre** [comment]
There is a strong possibility that it might be due to numpy arrays being too large, with BED files that are too big :

https://stackoverflow.com/questions/24406937/scikit-learn-joblib-bug-multiprocessing-pool-self-value-out-of-range-for-i-fo

**Comment by qferre** [comment]
I just tried with a random BED peak file with 1 million peaks, and it seemed to pass this step at least once (did not run it to completion since it it would take an insane amount of time)...

**Comment by qferre** [comment]
It could also be due to the massive number of jobs somehow overloading the memory : https://github.com/scikit-learn/scikit-learn/issues/3313

@dputhier , @Norisad : is Sacapus running 32 bit or 64 bit Anaconda ? How much memory is available on Sacapus ? Maybe we should add a provision to limit the number of jobs ?

Sorry, I know I'm grasping as straws, but I'm quite at a loss here.

**Comment by qferre** [comment]
Further nudge towards 32 bit being a problem : http://python.6.x6.nabble.com/test-gzip-test-tarfile-failure-om-AMD64-td1830323.html

Remember that `-2147483648 <= number <= 2147483647` means roughly `0.5 * -2**32 < number < 0.5 * 2**32`

**Comment by dputhier** [comment]
I dont think this is related to the number of jobs as they may not run concurrently on the nodes (we have reserved 12 of the 15 cpus).
I guess this is a 64 bit Anaconda...

**Comment by qferre** [comment]
This really looks like a problem that would arise with very large objects being multiprocessed : https://bugs.python.org/issue17560

The only problem is : **in this case, our BED files are small**. I even tried myself with a random BED of 1M region, and it worked.

It appears *something* is trying to pickle an extremely large object, and I don't know why.

**Comment by qferre** [comment]
Nori sent me the *exact same* files he used and I ran them locally. Everything worked, so we must assume this is an error with Sacapus. Maybe there is something weird with the memory management there.

TL;DR : could not reproduce the bug.

**Comment by dputhier** [comment]
Same version of Python exactly ? Same version of multiprocessing ?

Le jeu. 14 mars 2019 19:24, Quentin Ferré <notifications@github.com> a
écrit :

**Comment by qferre** [comment]
You were absolutely right to correct me. I apologize for my earlier rash statement : I indeed think we have different versions.

In my own pygtftk environment, I have Python 3.7.1, I am not sure that this is the version Nori used. I'll ask.

**Comment by dputhier** [comment]
I just ran almost the same jobs (H3K4me3) with my own files and everything worked perfectly on sacapus. The peak_file had 73703 lines and was compared against 17 files containing 31236,   19570,   29348,   24143,   21014,   20831,   28280,   19714,   26828,   18922,   25710,   20078,   23897,   29836,   31077,   21564 and  21564 line. 
No error.
 So it must be something related to file format (meaning that we should add additional tests) or file size or installation.

**Comment by dputhier** [comment]
I also tried the peak_file (73703 lines) against a file containing a merge of all peaks (413612 lines).
Worked like a charm.

**Comment by qferre** [comment]
*Holy sweet Mary mother of volcanoes*, what is wrong with the original files ? I checked them and a priori they are not different from any other BED file !

In any case, I don't think size is the issue, since I tried against a file of 1 million random peaks without error. 

My best guess is that there is something wrong with Nori's installation in conda, since we could not reproduce it. What do you think ?

**Comment by qferre** [comment]
Update according to dputhier qsub might be the problem.

**Comment by qferre** [comment]
We now believe the problem was due to some installation troubles Nori had.

@Norisad : can you tell me as soon as you've been able to redo the test ?

**Comment by qferre** [comment]
After a clean reinstall on Sacapus, it appears the error has disappeared ? Closing.
